app/models/site.rb

```diff
@@ -28,16 +28,57 @@ def user_fields
     UserField.order(:position).all
   end
 
-  def categories
-    @categories ||= begin
-      categories = Category
-        .includes(:uploaded_logo, :uploaded_background, :tags, :tag_groups)
-        .secured(@guardian)
+  CHANNEL = '/clear_site_cache'
+
+  def self.clear_cache
+    @categories_cache = nil
+    MessageBus.publish(CHANNEL, process: process_id) if !Rails.env.test?
+  end
+
+  def self.process_id
+    @process_id ||= SecureRandom.uuid
+  end
+
+  def self.after_fork
+    @process_id = nil
+    ensure_listen_for_changes
+  end
+
+  def self.ensure_listen_for_changes
+    unless @subscribed
+      MessageBus.subscribe(CHANNEL) do |message|
+        if message.data["process_id"] != process_id
+          @categories_cache = nil
+        end
+      end
+
+      @subscribed = true
+    end
+  end
+
+  def self.all_categories_cache
+    # Categories do not change often so there is no need for us to run the
+    # same query and spend time creating ActiveRecord objects for every requests.
+    #
+    # Do note that any new association added to the eager loading needs a
+    # corresponding ActiveRecord callback to clear the categories cache.
+    @categories_cache ||= begin
```

Comment:
Per process in memory cache instead of distributed cache because we don't want to be marshalling a ton of ActiveRecord objects and sending them flying through Redis.

Reply:
I'm not sure this is safe, because it means we'll be using the same active record objects in multiple threads.

Reply:
Oh, and it looks like this won't work with multisite.

Reply:
Hmm good point. This is starting to become tricky… I think I need to stop caching AR objects and revert it back to caching the serialized objects.